2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+---
+RFC: RFC
+Author: Dongbo Wang, Dan Travison
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine and Configuration
+Comments Due: March 22nd, 2018
+Plan to implement: Yes
+---
+
+# Support Experimental Features via Configuration
+
+This RFC is about adding the Experimental Feature support in PowerShell through the PowerShell configuration.
+
+The Experimental Feature support in PowerShell is to provide a mechanism for experimental features to coexist
+with existing stable features in PowerShell or PowerShell modules.
+An experimental feature would not be enabled
+unless a user opt in by properly configuring the `powershell.config.json` file.
+Experimental features should not affect any stable features
+and not present any side effects
+when they are all disabled.
+
+## Motivation
+
+This is a highly demanded feature which we believe will vastly improve the productivity of
+PowerShell developers on experimenting or iterating new features in PowerShell and PowerShell modules.
+
+The Experimental Feature support would make it easy for users to try out features
+that are in early or intermediate stages of development,
+and thus, allow the developers to get early feedback
+and make adjustments to the design or implementation based on the feedback.
+
+For users that do not care about the new features,
+they can continue to enjoy the bug fixes and other improvements from the new releases
+without worrying about regressions, issues, or side effects caused by the new features.
+
+For example, PowerShell team plans to rewrite the `FileSystemProvider` in future
+to reduce the code complexity and improve the performance.
+During the development,
+the `FileSystemProviderV2` can coexist with the existing one
+and be exposed to the user as an experimental feature,
+so that PowerShell team can get early feedback and issue reports.
```

Comment:
RFC updated to add `-Enabled` and `-Disabled` switches.

Reply:
Without this we have to ask, for ex., in every file system Issue - Did you enable FileSystemProviderV2?

Reply:
We can add another environment section in the issue template called `Get-ExperimentalFeatures -Enabled`.
If there is really a demand, we can add information about experimental features to `$PSVertionTable`, but I don't want to add it to the RFC.

Reply:
Rather than modify $PSVersionTable or depend on a this command specific to this feature, I think it may be beneficial to consider a an RFC for a generalized PowerShell support cmdlet that provides common environment details asked for by the PowerShell project and many other open source projects for PowerShell. It would be convenient for us and the community. I envision it as a kind of `phpinfo()` for PowerShell. Adding the command to the template is a good stop-gap though and if anyone else sees value in such a command for PowerShell I will consider an RFC.